Switch to TextIndicator for WebKitLegacy.

megangardner opened the pull request:
#### 9d0d94b77ec1f330081dbb0c502c1eae35ffaa67
<pre>
Switch to TextIndicator for WebKitLegacy.
<a href="https://bugs.webkit.org/show_bug.cgi?id=296287">https://bugs.webkit.org/show_bug.cgi?id=296287</a>
<a href="https://rdar.apple.com/156337126">rdar://156337126</a>

Reviewed by Abrar Rahman Protyasha.

Continuing the removal of TextIndicator Data.

* Source/WebCore/page/TextIndicator.h:
(WebCore::TextIndicator::estimatedBackgroundColor const):
(WebCore::TextIndicator::options const):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebUITextIndicatorData initWithImage:textIndicator:scale:]):
(-[WebView _startDrag:]):
(-[WebView _didConcludeEditDrag]):
(-[WebUITextIndicatorData initWithImage:textIndicatorData:scale:]): Deleted.
* Source/WebKitLegacy/mac/WebView/WebViewInternal.h:

Canonical link: <a href="https://commits.webkit.org/297729@main">https://commits.webkit.org/297729@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/92d314ad762fb7e701772f1f54fb0491269104ae

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112620 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32352 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22830 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118819 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63090 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114582 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33004 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40915 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85713 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/36335 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115567 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26340 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101318 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66019 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25636 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19455 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62578 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95728 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19530 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122040 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39694 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29568 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94582 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40077 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97556 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94323 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24087 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39440 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17246 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/35782 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39582 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45070 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39221 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42554 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40960 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKitLegacy/mac/WebView/WebView.mm:
```diff
@@ -827,31 +827,31 @@ - (void)dealloc
+        _contentImage = [PAL::allocUIImageInstance() initWithCGImage:indicator->contentImage()->nativeImage()->platformImage().get() scale:scale orientation:UIImageOrientationUp];
+
+    if (indicatorData->contentImageWithoutSelection()) {
+        auto nativeImage = indicator->contentImageWithoutSelection()->nativeImage();
```
These would all require `indicator.` instead of `indicator->`, right? Since it's a reference

aprotyas replied:
Oh, maybe you meant to pass a `RefPtr<TextIndicator>` to the method instead

aprotyas commented on Source/WebKitLegacy/mac/WebView/WebView.mm:
```diff
@@ -2054,7 +2054,7 @@ - (void)_didConcludeEditDrag
         return;
     if (auto range = frame->selection().selection().toNormalizedRange()) {
         if (auto textIndicator = WebCore::TextIndicator::createWithRange(*range, defaultEditDragTextIndicatorOptions, WebCore::TextIndicatorPresentationTransition::None, WebCore::FloatSize()))
```
Drive-by while you're in here!
```suggestion
        if (RefPtr textIndicator = WebCore::TextIndicator::createWithRange(*range, defaultEditDragTextIndicatorOptions, WebCore::TextIndicatorPresentationTransition::None, WebCore::FloatSize()))
```

aprotyas commented on Source/WebKitLegacy/mac/WebView/WebViewInternal.h:
```diff
@@ -126,7 +126,7 @@ WebLayoutMilestones kitLayoutMilestones(OptionSet<WebCore::LayoutMilestone>);
 #if PLATFORM(IOS_FAMILY) && ENABLE(DRAG_SUPPORT) && defined(__cplusplus)
 @interface WebUITextIndicatorData (WebUITextIndicatorInternal)
-- (WebUITextIndicatorData *)initWithImage:(CGImageRef)image textIndicatorData:(const WebCore::TextIndicatorData&)indicatorData scale:(CGFloat)scale;
+- (WebUITextIndicatorData *)initWithImage:(CGImageRef)image textIndicatorData:(const WebCore::TextIndicator&)indicator scale:(CGFloat)scale;
```
I think this will have to be a RefPtr
```suggestion
- (WebUITextIndicatorData *)initWithImage:(CGImageRef)image textIndicatorData:(RefPtr<WebCore::TextIndicator>&&)indicator scale:(CGFloat)scale;
```

and then you can change your call sites to

```objc
adoptNS([[WebUITextIndicatorData alloc] initWithImage:nil textIndicator:WTFMove(textIndicator) scale:page->deviceScaleFactor()]);
```

aprotyas reviewed (commented):
Consider `const&` --> `RefPtr&&`?

pxlcoder commented on Source/WebCore/page/TextIndicator.h:
```diff
@@ -173,6 +173,10 @@ class TextIndicator : public RefCounted<TextIndicator> {
     WEBCORE_EXPORT bool wantsBounce() const;
     WEBCORE_EXPORT bool wantsManualAnimation() const;
 
+
```
There's an extra newline here.

megangardner committed:
Switch to TextIndicator for WebKitLegacy.
https://bugs.webkit.org/show_bug.cgi?id=296287
rdar://156337126

Reviewed by Abrar Rahman Protyasha.

Continuing the removal of TextIndicator Data.

* Source/WebCore/page/TextIndicator.h:
(WebCore::TextIndicator::estimatedBackgroundColor const):
(WebCore::TextIndicator::options const):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebUITextIndicatorData initWithImage:textIndicator:scale:]):
(-[WebView _startDrag:]):
(-[WebView _didConcludeEditDrag]):
(-[WebUITextIndicatorData initWithImage:textIndicatorData:scale:]): Deleted.
* Source/WebKitLegacy/mac/WebView/WebViewInternal.h:

Canonical link: https://commits.webkit.org/297729@main
